src/libai/nn/NeuralNetwork.java

```diff
@@ -67,21 +80,30 @@ public void setProgressBar(ProgressDisplay pb) {
 	 * @param length	How many patterns will be used.
 	 * @param minerror	The minimal error expected.
 	 */
-	public abstract void train(Matrix[] patterns, Matrix[] answers, double alpha, int epochs, int offset, int length, double minerror);
+	public void train(Matrix[] patterns, Matrix[] answers, double alpha, int epochs, int offset, int length, double minerror) {
```

Comment:
This kind of breaks the idea of the abstraction of the NN. Yes, most validations are common to most NN, but for instance, the minerror depends entirely on the error function used, most error functions are non-negative functions, but if a new NN appears that uses a function that maximizes instead of minimized then the expected minerror might be negative. Maximization problems can be rewritten to minimization ones as `max(f(x)) = -min(-f(x))`